swiftnav/src/time/gnss.rs

```diff
@@ -0,0 +1,816 @@
+        /* the UTC offset takes effect exactly 1 second after the start of
+         * the (positive) leap second event */
+        let dt = self.diff(&params.t_lse());
+
+        /* True only when self is during the leap second event */
+        (0.0..1.0).contains(&dt)
+    }
+
+    /// Checks to see if this point in time is a UTC leap second event using the
+    /// hardcoded list of leap seconds
+    ///
+    /// # ⚠️  🦘  ⏱  ⚠️  - Leap Seconds
+    ///
+    /// The hard coded list of leap seconds will get out of date, it is
+    /// preferable to use [`GpsTime::is_leap_second_event()`] with the newest
+    /// set of UTC parameters
+    pub fn is_leap_second_event_hardcoded(&self) -> bool {
+        for (t_leap, _offset) in UTC_LEAPS.iter().rev() {
+            let dt = self.diff(t_leap);
+
+            if dt > 1.0 {
+                /* time is past the last known leap second event */
+                return false;
+            }
+            if (0.0..1.0).contains(&dt) {
+                /* time is during the leap second event */
+                return true;
+            }
+        }
+
+        /* time is before the first known leap second event */
+        false
+    }
+
+    /// Converts the GPS time into a [`MJD`] (modified julian date)
+    pub fn to_mjd(self, utc_params: &UtcParams) -> MJD {
+        self.to_utc(utc_params).to_mjd()
+    }
+
+    /// Converts the GPS time into a [`MJD`] (modified julian date) using a hard
+    /// coded list of leap seconds
+    ///
+    /// # ⚠️  🦘  ⏱  ⚠️  - Leap Seconds
+    ///
+    /// The hard coded list of leap seconds will get out of date, it is
+    /// preferable to use [`GpsTime::to_mjd()`] with the newest
+    /// set of UTC parameters
+    pub fn to_mjd_hardcoded(self) -> MJD {
+        self.to_utc_hardcoded().to_mjd()
+    }
+
+    /// Gets the GPS time of the nearest solution epoch
+    pub fn round_to_epoch(&self, soln_freq: f64) -> GpsTime {
+        let rounded_tow = (self.tow * soln_freq).round() / soln_freq;
+        let mut rounded_time = Self::new_unchecked(self.wn, rounded_tow);
+        /* handle case where rounding caused tow roll-over */
+        rounded_time.normalize();
+        rounded_time
+    }
+
+    /// Gets the GPS time of the previous solution epoch
+    pub fn floor_to_epoch(&self, soln_freq: f64) -> GpsTime {
+        /* round the time-of-week */
+        let rounded_tow = (self.tow * soln_freq).floor() / soln_freq;
+        let mut rounded_time = GpsTime::new_unchecked(self.wn, rounded_tow);
+        /* handle case where rounding caused tow roll-over */
+        rounded_time.normalize();
+        rounded_time
+    }
+
+    /// Converts the GPS time into Galileo time
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if the GPS time is before the start of Galileo
+    /// time, i.e. [`GAL_TIME_START`]
+    pub fn to_gal(self) -> GalTime {
+        assert!(self.is_valid());
+        assert!(self >= GAL_TIME_START);
```

Comment:
Same question here for most of this, I would expect returning an error will be better. We can let the caller decide if they would like to panic or not rather than forcing them to do so here.

The only time i'd encourage panics is if there is truly an unrecoverable error that occurs.